detectron2/modeling/meta_arch/fcos.py

```diff
@@ -96,77 +102,73 @@ def forward_training(self, images, features, predictions, gt_instances):
+        gt_labels = []
+        matched_gt_boxes = []
         for gt_per_image in gt_instances:
-            gt_centers = gt_per_image.gt_boxes.get_centers()  # Nx2
-            # FCOS with center sampling: anchor point must be close enough to gt center.
-            pairwise_match = (anchor_centers[:, None, :] - gt_centers[None, :, :]).abs_().max(
-                dim=2
-            ).values < self.center_sampling_radius * anchor_sizes[:, None]
+            gt_centers = gt_per_image.gt_boxes.get_centers()
+
+            # FCOS with center sampling: anchor point must be close enough to
+            # ground-truth box center.
+            center_dists = (anchor_centers[None, :, :] - gt_centers[:, None, :]).abs_()
+            sampling_regions = self.center_sampling_radius * anchor_sizes[None, :]
+
+            match_quality_matrix = center_dists.max(dim=2).values < sampling_regions  # (M, R)
+
             pairwise_dist = pairwise_point_box_distance(anchor_centers, gt_per_image.gt_boxes)
+            pairwise_dist = pairwise_dist.permute(1, 0, 2)  # (M, R, 4)
 
-            # The original FCOS anchor matching rule: anchor point must be inside gt
-            pairwise_match &= pairwise_dist.min(dim=2).values > 0
+            # The original FCOS anchor matching rule: anchor point must be inside GT.
+            match_quality_matrix &= pairwise_dist.min(dim=2).values > 0
 
             # Multilevel anchor matching in FCOS: each anchor is only responsible
             # for certain scale range.
             pairwise_dist = pairwise_dist.max(dim=2).values
-            pairwise_match &= (pairwise_dist > lower_bound[:, None]) & (
-                pairwise_dist < upper_bound[:, None]
+            match_quality_matrix &= (pairwise_dist > lower_bound[None, :]) & (
+                pairwise_dist < upper_bound[None, :]
             )
+            # Match the GT box with minimum area, if there are multiple GT matches.
+            gt_areas = gt_per_image.gt_boxes.area()  # (M, )
 
-            # Match the GT box with minimum area, if there are multiple GT matches
-            gt_areas = gt_per_image.gt_boxes.area()  # N
-            pairwise_match = pairwise_match.to(torch.float32) * (1e8 - gt_areas[None, :])
-            min_values, matched_idx = pairwise_match.max(dim=1)  # R, per-anchor match
-            matched_idx[min_values < 1e-5] = -1  # Unmatched anchors are assigned -1
-
-            matched_indices.append(matched_idx)
-        return matched_indices
-
-    @torch.no_grad()
-    def label_anchors(self, anchors, gt_instances):
-        """
-        Same interface as :meth:`RetinaNet.label_anchors`, but implemented with FCOS
-        anchor matching rule.
+            # Get matches and their labels using match quality matrix.
+            match_quality_matrix = match_quality_matrix.to(torch.float32) * (
+                1e8 - gt_areas[:, None]
+            )
+            matched_idxs, anchor_labels = self.anchor_matcher(match_quality_matrix)
```

Comment:
Though in general I like showing that RetinaNet and FCOS are very similar, I hope to not achieve this by complicating FCOS. After all it does deserve the credit of simplicity. 

Personally I think the introduction of anchor/box_transform is not complicating FCOS because it does make many RetinaNet code reusable, and make this implementation easier to understand than its original code IMHO. But introducing an extra concept of "anchor matcher" that can be achieved with a trivial one-liner is unnecessarily complicating the code.

Perhaps it's better to keep the original two lines here, but with a comment saying it is equivalent to applying `Matcher(1e-5, [0, 1])`. This way we keep the simplicity of FCOS while also drawing the connection to RetinaNet.